crates/core/tedge_agent/src/http_rest.rs

```diff
@@ -0,0 +1,371 @@
+        if clean_path.starts_with(&self.file_transfer_dir) {
+            Ok(clean_path)
+        } else {
+            Err(FileTransferError::InvalidURI {
+                value: clean_path.to_string_lossy().to_string(),
+            })
+        }
+    }
+}
+
+mod uri_utils {
+    // below are a set of utility function used for working with
+    // URI coming from GET, PUT and DELETE requests to the thin-edge
+    // device.
+    use std::path::PathBuf;
+
+    pub fn separate_path_and_file_name(input: PathBuf) -> Option<(PathBuf, String)> {
+        let input_as_str = input.to_str()?;
+        let (relative_path, file_name) = input_as_str.rsplit_once('/')?;
+
+        let relative_path = PathBuf::from(relative_path);
+        Some((relative_path, file_name.into()))
+    }
+}
+
+async fn put(
+    mut request: Request<Body>,
+    file_transfer: &HttpConfig,
+) -> Result<Response<Body>, FileTransferError> {
+    let full_path = file_transfer.local_path_for_uri(request.uri().to_string())?;
+
+    let mut response = Response::new(Body::empty());
+
+    if let Some((relative_path, file_name)) = uri_utils::separate_path_and_file_name(full_path) {
+        let root_path = file_transfer.file_transfer_dir.clone();
+        let directories_path = root_path.join(relative_path);
+
+        if let Err(_err) = create_directories(&directories_path) {
+            *response.status_mut() = hyper::StatusCode::FORBIDDEN;
+        }
+
+        let full_path = directories_path.join(file_name);
+
+        match stream_request_body_to_path(&full_path, request.body_mut()).await {
+            Ok(()) => {
+                *response.status_mut() = hyper::StatusCode::CREATED;
+            }
+            Err(_err) => {
+                *response.status_mut() = hyper::StatusCode::FORBIDDEN;
+            }
+        }
+    } else {
+        *response.status_mut() = hyper::StatusCode::FORBIDDEN;
+    }
+    Ok(response)
+}
+
+async fn get(
+    request: Request<Body>,
+    file_transfer: &HttpConfig,
+) -> Result<Response<Body>, FileTransferError> {
+    let full_path = file_transfer.local_path_for_uri(request.uri().to_string())?;
+
+    if !full_path.exists() || full_path.is_dir() {
+        let mut response = Response::new(Body::empty());
+        *response.status_mut() = hyper::StatusCode::NOT_FOUND;
+        return Ok(response);
+    }
+
+    let mut file = tokio::fs::File::open(full_path).await?;
+
+    let mut contents = vec![];
+    file.read_to_end(&mut contents).await?;
+
+    let output = String::from_utf8(contents)?;
+
+    Ok(Response::new(Body::from(output)))
+}
+
+async fn delete(
+    request: Request<Body>,
+    file_transfer: &HttpConfig,
+) -> Result<Response<Body>, FileTransferError> {
+    let full_path = file_transfer.local_path_for_uri(request.uri().to_string())?;
+
+    let mut response = Response::new(Body::empty());
+
+    if !full_path.exists() {
+        *response.status_mut() = hyper::StatusCode::INTERNAL_SERVER_ERROR;
```

Comment:
Isn't it better to return an OK 200 code even if the file does not exist on a DELETE request? So that the DELETE request remains idempotent on back-to-back calls as well? I thought that's what we had agreed on, while discussing the HTTP status codes.

Reply:
Indeed, it would be better to return OK 200 when a file to be removed doesn't exist.